presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -806,7 +859,50 @@ public void dropPartition(String databaseName, String tableName, List<String> pa
     }
 
     @Override
-    public void alterPartition(String databaseName, String tableName, Partition partition)
+    public void alterPartition(String databaseName, String tableName, PartitionWithStatistics partitionWithStatistics)
+    {
+        Partition oldPartition = getPartition(databaseName, tableName, partitionWithStatistics.getPartition().getValues())
+                .orElseThrow(() -> new PartitionNotFoundException(new SchemaTableName(databaseName, tableName), partitionWithStatistics.getPartition().getValues()));
+        Table table = getTable(databaseName, tableName)
+                .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(databaseName, tableName)));
+        List<String> partitionColumns = table.getPartitionKeys().stream()
+                .map(FieldSchema::getName)
+                .collect(toImmutableList());
+        String partitionName = makePartName(partitionColumns, oldPartition.getValues());
+        List<String> dataColumns = table.getSd().getCols().stream()
+                .map(FieldSchema::getName)
+                .collect(toImmutableList());
+        List<ColumnStatisticsObj> oldStatistics = getMetastorePartitionColumnStatistics(table.getDbName(), table.getTableName(), ImmutableSet.of(partitionName), dataColumns)
+                .getOrDefault(partitionName, ImmutableList.of());
+
+        alterMetastorePartition(databaseName, tableName, toMetastoreApiPartition(partitionWithStatistics));
+
+        try {
+            // remove statistics for the columns that will not be overwritten
+            Set<String> columnStatisticsForRemoval = oldStatistics.stream()
+                    .map(ColumnStatisticsObj::getColName)
+                    .collect(toSet());
+            columnStatisticsForRemoval.removeAll(partitionWithStatistics.getStatistics().getColumnStatistics().keySet());
+            for (String columnName : columnStatisticsForRemoval) {
+                deletePartitionColumnStatistics(databaseName, tableName, partitionName, columnName);
```

Comment:
It is not acceptable to have `O(#partition * #column) ` calls. Any alternatives?